File: packages/rossetta-client/src/obfuscation.ts
/**
 * Google-style BatchExecute/Wrb.fr obfuscation layer
 * Transforms encrypted payloads into Google API-like format for enhanced obfuscation
 * 
 * This module implements a sophisticated obfuscation scheme that mimics Google's 
 * BatchExecute API responses, making it harder to identify encrypted data within 
 * network traffic.
 */

/**
 * Session state for Google-style requests
 */
interface SessionState {
  sid: string;
  reqId: number;
  at: string;
}

// Global session state
let sessionState: SessionState | null = null;

/**
 * Generate a random session ID (f.sid format)
 */
function generateSessionId(): string {
  // Generate a 19-digit numeric string like Google's f.sid
  const chars = '0123456789';
  let result = '';
  for (let i = 0; i < 19; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

/**
 * Generate an action token (CSRF-like token)
 */
function generateActionToken(): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let result = 'AMsTio';
  for (let i = 0; i < 20; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  // Add timestamp component
  result += ':' + Date.now();
  return result;
}

/**
 * Initialize or get session state
 */
export function getSessionState(): SessionState {
  if (!sessionState) {
    sessionState = {
      sid: generateSessionId(),
      reqId: Math.floor(Math.random() * 100000),
      at: generateActionToken(),
    };
  }
  return sessionState;
}

/**
 * Reset session state (useful for testing)
 */
export function resetSessionState(): void {
  sessionState = null;
}

/**
 * Increment request ID
 */
export function incrementReqId(): number {
  const state = getSessionState();
  state.reqId += Math.floor(Math.random() * 1000) + 100;
  return state.reqId;
}

/**
 * Generate Google-style request headers
 */
export function generateRequestHeaders(): Record<string, string> {
  const state = getSessionState();
  return {
    'f.sid': state.sid,
    'bl': 'boq_rossetta-frontend-ui_' + new Date().toISOString().slice(0, 10).replace(/-/g, '') + '.01_p0',
    'hl': 'en-GB',
    'gl': 'GB',
    'soc-app': '162',
    'soc-platform': '1',
    'soc-device': '1',
    '_reqid': String(incrementReqId()),
    'rt': 'c',
  };
}

/**
 * Encode encrypted data into Google BatchExecute-style format
 * 
 * Transforms { ciphertext, iv } into a nested array structure
 * that mimics Google's internal API format
 */
export function encodeToGoogleFormat(ciphertext: string, iv: string): {
  formData: Record<string, string>;
  headers: Record<string, string>;
} {
  const state = getSessionState();
  const timestamp = Date.now();
  
  // Create nested array structure similar to Google's format
  // The structure includes: wrb.fr wrapper, encrypted data, metadata
  const innerPayload = [
    null,
    [
      [timestamp, Math.floor(Math.random() * 1000000000), Math.floor(Math.random() * 10000000000)],
      null,
      null,
      null,
      null,
      [[1]]
    ],
    0,
    ciphertext,  // Our encrypted ciphertext
    iv,          // Our IV
  ];
  
  const wrbFrPayload = [
    ['wrb.fr', null, JSON.stringify(innerPayload)]
  ];
  
  // The f.req field contains the nested array structure
  const fReq = JSON.stringify([
    null,
    JSON.stringify(wrbFrPayload)
  ]);
  
  return {
    formData: {
      'f.req': fReq,
      'at': state.at,
    },
    headers: generateRequestHeaders(),
  };
}

/**
 * Decode Google BatchExecute-style format back to original encrypted data
 * 
 * Extracts { ciphertext, iv } from the nested array structure
 */
export function decodeFromGoogleFormat(body: string): { ciphertext: string; iv: string } | null {
  try {
    // Parse the outer structure
    const parsed = JSON.parse(body);
    
    if (!Array.isArray(parsed) || parsed.length < 2) {
      return null;
    }
    
    // Parse the inner wrapper string
    const innerStr = parsed[1];
    const inner = JSON.parse(innerStr);
    
    if (!Array.isArray(inner) || inner.length === 0) {
      return null;
    }
    
    // Get the wrb.fr payload
    const wrbFr = inner[0];
    if (!Array.isArray(wrbFr) || wrbFr.length < 3 || wrbFr[0] !== 'wrb.fr') {
      return null;
    }
    
    // Parse the innermost payload
    const payloadStr = wrbFr[2];
    const payload = JSON.parse(payloadStr);
    
    if (!Array.isArray(payload) || payload.length < 5) {
      return null;
    }
    
    // Extract ciphertext and iv from our positions
    const ciphertext = payload[3];
    const iv = payload[4];
    
    if (typeof ciphertext !== 'string' || typeof iv !== 'string') {
      return null;
    }
    
    return { ciphertext, iv };
  } catch {
    return null;
  }
}

/**
 * Reverse Unicode escaping applied during encoding
 */
function unescapeUnicode(s: string): string {
  const replacements: Record<string, string> = {
    '\\u003d': '=',
    '\\u003c': '<',
    '\\u003e': '>',
    '\\u0026': '&',
  };
  let result = s;
  for (const [escape, char] of Object.entries(replacements)) {
    result = result.split(escape).join(char);
  }
  return result;
}

// Marker prefix used to identify scattered data during reassembly
const DATA_SCATTER_MARKER = 'RX1';

/**
 * Reassemble scattered data parts by joining them and removing the marker prefix.
 */
function reassembleScatteredData(parts: string[]): string | null {
  if (parts.length === 0) {
    return null;
  }
  
  const combined = parts.join('');
  
  // Check for our marker prefix
  if (combined.startsWith(DATA_SCATTER_MARKER)) {
    return combined.slice(DATA_SCATTER_MARKER.length);  // Remove marker
  }
  
  return null;
}

/**
 * Parse Google's multi-segment response format
 * 
 * Format: <length>\n<json_segment><length>\n<json_segment>...
 */
function parseGoogleSegments(body: string): unknown[] {
  const segments: unknown[] = [];
  let remaining = body.trim();
  
  while (remaining.length > 0) {
    // Find the length prefix
    const newlineIdx = remaining.indexOf('\n');
    if (newlineIdx === -1) {
      break;
    }
    
    const lengthStr = remaining.slice(0, newlineIdx);
    if (!/^\d+$/.test(lengthStr)) {
      break;
    }
    
    const segmentLength = parseInt(lengthStr, 10);
    const segmentStart = newlineIdx + 1;
    const segmentEnd = segmentStart + segmentLength;
    
    let segmentStr: string;
    if (segmentEnd > remaining.length) {
      // Partial segment - try to parse what we have
      segmentStr = remaining.slice(segmentStart);
    } else {
      segmentStr = remaining.slice(segmentStart, segmentEnd);
    }
    
    try {
      const segment = JSON.parse(segmentStr);
      segments.push(segment);
    } catch {
      // Skip invalid segments
    }
    
    remaining = segmentEnd <= remaining.length ? remaining.slice(segmentEnd) : '';
  }
  
  return segments;
}

/**
 * Encode response in Google BatchExecute-style format
 * 
 * Creates a response format that mimics Google's API responses
 */
export function encodeResponseToGoogleFormat(ciphertext: string, iv: string): string {
  const timestamp = Date.now();
  
  // Create the response structure
  const responsePayload = [
    null,
    [
      [timestamp, Math.floor(Math.random() * 1000000000), Math.floor(Math.random() * 10000000000)],
      null,
      null,
      null,
      null,
      [[2]]
    ],
    1,
    ciphertext,
    iv,
  ];
  
  // Wrap in wrb.fr format
  const wrbFrResponse = [
    ['wrb.fr', null, JSON.stringify(responsePayload)]
  ];
  
  // Add length prefix like Google does
  const responseStr = JSON.stringify(wrbFrResponse);
  const lengthPrefix = String(responseStr.length);
  
  return lengthPrefix + '\n' + responseStr;
}

/**
 * Decode Google BatchExecute-style response to original encrypted data
 * 
 * Handles both the new sophisticated scattered format and the legacy format
 */
export function decodeResponseFromGoogleFormat(body: string): { ciphertext: string; iv: string } | null {
  try {
    // Parse the multi-segment response
    const segments = parseGoogleSegments(body);
    
    if (segments.length === 0) {
      return null;
    }
    
    // Find the wrb.fr segment
    let wrbFrData: unknown[] | null = null;
    for (const segment of segments) {
      if (Array.isArray(segment) && segment.length > 0) {
        const inner = segment[0];
        if (Array.isArray(inner) && inner.length >= 3 && inner[0] === 'wrb.fr') {
          wrbFrData = inner;
          break;
        }
      }
    }
    
    if (!wrbFrData) {
      return null;
    }
    
    // Parse the inner payload
    const innerStr = wrbFrData[2] as string;
    const inner = JSON.parse(innerStr);
    
    if (!Array.isArray(inner)) {
      return null;
    }
    
    // Extract scattered ciphertext and IV parts from the nested structure
    // Structure documentation (matching Python encode_response_to_google_format):
    // - inner[4][0][0][1][1]: ciphertext part 1 (in resource metadata)
    // - inner[4][0][0][6][5]: IV part 1 (in multi-carrier info)  
    // - inner[7]: ciphertext part 2 (standalone slot)
    // - inner[8][2]: ciphertext part 3 (in timestamp array)
    // - inner[9]: IV part 2 (standalone slot)
    const ctParts: string[] = [];
    const ivParts: string[] = [];
    
    // Ciphertext part 1: inner[4][0][0][1][1] (in resource metadata)
    try {
      const part = (inner as unknown[])[4];
      if (Array.isArray(part) && part[0] && Array.isArray(part[0][0]) && part[0][0][1]) {
        const val = part[0][0][1][1];
        if (typeof val === 'string') {
          ctParts.push(unescapeUnicode(val));
        }
      }
    } catch {
      // Part not found
    }
    
    // Ciphertext part 2: inner[7] (standalone slot)
    try {
      const val = (inner as unknown[])[7];
      if (typeof val === 'string') {
        ctParts.push(unescapeUnicode(val));
      }
    } catch {
      // Part not found
    }
    
    // Ciphertext part 3: inner[8][2] (in timestamp array)
    try {
      const arr = (inner as unknown[])[8];
      if (Array.isArray(arr) && typeof arr[2] === 'string') {
        ctParts.push(unescapeUnicode(arr[2]));
      }
    } catch {
      // Part not found
    }
    
    // IV part 1: inner[4][0][0][6][5] (in multi-carrier info)
    try {
      const part = (inner as unknown[])[4];
      if (Array.isArray(part) && part[0] && Array.isArray(part[0][0]) && part[0][0][6]) {
        const val = part[0][0][6][5];
        if (typeof val === 'string') {
          ivParts.push(unescapeUnicode(val));
        }
      }
    } catch {
      // Part not found
    }
    
    // IV part 2: inner[9] (standalone slot)
    try {
      const val = (inner as unknown[])[9];
      if (typeof val === 'string') {
        ivParts.push(unescapeUnicode(val));
      }
    } catch {
      // Part not found
    }
    
    // Reassemble the scattered parts back into complete data
    if (ctParts.length > 0 && ivParts.length > 0) {
      const ciphertext = reassembleScatteredData(ctParts);
      const iv = reassembleScatteredData(ivParts);
      
      if (ciphertext && iv) {
        return { ciphertext, iv };
      }
    }
    
    // Fallback to legacy format (positions 3 and 4)
    const legacyCiphertext = inner[3];
    const legacyIv = inner[4];
    
    if (typeof legacyCiphertext === 'string' && typeof legacyIv === 'string') {
      return { ciphertext: legacyCiphertext, iv: legacyIv };
    }
    
    return null;
  } catch {
    return null;
  }
}

/**
 * Convert form data to URL-encoded format
 */
export function toFormUrlEncoded(data: Record<string, string>): string {
  return Object.entries(data)
    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
    .join('&');
}

File: packages/rossetta-fastapi/rossetta_fastapi/obfuscation.py
"""
Google-style BatchExecute/Wrb.fr obfuscation layer
Transforms encrypted payloads into Google API-like format for enhanced obfuscation

This module implements a sophisticated obfuscation scheme that mimics Google's 
BatchExecute API responses, making it harder to identify encrypted data within 
network traffic.
"""

import base64
import json
import secrets
import time
import urllib.parse
from typing import Optional, Tuple, Dict, Any, List


class SessionState:
    """Session state for Google-style requests"""
    
    def __init__(self):
        self.sid = self._generate_session_id()
        self.req_id = secrets.randbelow(100000)
        self.at = self._generate_action_token()
    
    @staticmethod
    def _generate_session_id() -> str:
        """Generate a random session ID (f.sid format) - 19-digit numeric string"""
        return ''.join(str(secrets.randbelow(10)) for _ in range(19))
    
    @staticmethod
    def _generate_action_token() -> str:
        """Generate an action token (CSRF-like token)"""
        chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
        result = 'AMsTio'
        result += ''.join(secrets.choice(chars) for _ in range(20))
        result += ':' + str(int(time.time() * 1000))
        return result
    
    def increment_req_id(self) -> int:
        """Increment request ID"""
        self.req_id += secrets.randbelow(1001) + 100
        return self.req_id


# Global session state
_session_state: Optional[SessionState] = None


def get_session_state() -> SessionState:
    """Initialize or get session state"""
    global _session_state
    if _session_state is None:
        _session_state = SessionState()
    return _session_state


def reset_session_state() -> None:
    """Reset session state (useful for testing)"""
    global _session_state
    _session_state = None


def _generate_google_id() -> str:
    """Generate a random Google-style identifier"""
    chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_'
    length = secrets.randbelow(8) + 16  # 16-24 chars
    return ''.join(secrets.choice(chars) for _ in range(length))


def _generate_resource_path() -> str:
    """Generate a random Google-style resource path like /m/0_5xc"""
    hex_chars = '0123456789abcdef'
    suffix_len = secrets.randbelow(4) + 3
    suffix = ''.join(secrets.choice(hex_chars) for _ in range(suffix_len))
    return f"/m/{suffix}"


def _apply_unicode_escaping(s: str) -> str:
    """Apply selective Unicode escaping to make output look more like Google's"""
    # Replace some characters with their Unicode escape sequences
    replacements = {
        '=': '\\u003d',
        '<': '\\u003c',
        '>': '\\u003e',
        '&': '\\u0026',
    }
    result = s
    for char, escape in replacements.items():
        result = result.replace(char, escape)
    return result


# Marker prefix for scattered data identification during reassembly
DATA_SCATTER_MARKER = "RX1"


def _split_data_for_scatter(data: str, num_parts: int = 3) -> List[str]:
    """
    Split data into multiple parts for scattering across the response.
    
    The data is prefixed with DATA_SCATTER_MARKER to enable reliable reassembly.
    Parts are split roughly equally in size.
    
    Args:
        data: The data to split
        num_parts: Number of parts to split into
        
    Returns:
        List of data parts, each containing a portion of the marked data
    """
    if num_parts < 2:
        return [data]
    
    # Add a prefix marker to identify our data format during reassembly
    marked_data = f"{DATA_SCATTER_MARKER}{data}"
    
    # Calculate part sizes (roughly equal)
    total_len = len(marked_data)
    base_size = total_len // num_parts
    remainder = total_len % num_parts
    
    parts = []
    offset = 0
    for i in range(num_parts):
        # Add one extra char to first 'remainder' parts
        size = base_size + (1 if i < remainder else 0)
        parts.append(marked_data[offset:offset + size])
        offset += size
    
    return parts


def _reassemble_scattered_data(parts: List[str]) -> Optional[str]:
    """Reassemble scattered data parts by joining them and removing the marker prefix."""
    if not parts:
        return None
    
    combined = ''.join(parts)
    
    # Check for our marker prefix
    if combined.startswith(DATA_SCATTER_MARKER):
        return combined[len(DATA_SCATTER_MARKER):]  # Remove marker
    
    return None


def _safe_get_part(parts: List[str], index: int) -> Optional[str]:
    """Safely get a part from the list, returning None if index is out of bounds."""
    return parts[index] if index < len(parts) else None


def decode_from_google_format(body: str) -> Optional[Tuple[str, str]]:
    """
    Decode Google BatchExecute-style format back to original encrypted data
    
    Extracts (ciphertext, iv) from the nested array structure.
    Supports both legacy format (positions 3,4) and new scattered format.
    """
    try:
        # URL decode if needed
        if '%5B' in body or '%5D' in body:
            body = urllib.parse.unquote(body)
        
        # Parse the outer structure
        parsed = json.loads(body)
        
        if not isinstance(parsed, list) or len(parsed) < 2:
            return None
        
        # Parse the inner wrapper string
        inner_str = parsed[1]
        inner = json.loads(inner_str)
        
        if not isinstance(inner, list) or len(inner) == 0:
            return None
        
        # Get the wrb.fr payload
        wrb_fr = inner[0]
        if not isinstance(wrb_fr, list) or len(wrb_fr) < 3 or wrb_fr[0] != 'wrb.fr':
            return None
        
        # Parse the innermost payload
        payload_str = wrb_fr[2]
        payload = json.loads(payload_str)
        
        if not isinstance(payload, list) or len(payload) < 5:
            return None
        
        # Extract ciphertext and iv from our positions
        ciphertext = payload[3]
        iv = payload[4]
        
        if not isinstance(ciphertext, str) or not isinstance(iv, str):
            return None
        
        return ciphertext, iv
    except (json.JSONDecodeError, TypeError, KeyError, IndexError):
        return None


def decode_from_form_data(form_data: Dict[str, str]) -> Optional[Tuple[str, str]]:
    """
    Decode encrypted data from form data format
    
    Extracts (ciphertext, iv) from f.req field in form data
    """
    try:
        f_req = form_data.get('f.req')
        if not f_req:
            return None
        
        return decode_from_google_format(f_req)
    except Exception:
        return None


def encode_response_to_google_format(ciphertext: str, iv: str) -> str:
    """
    Encode response in Google BatchExecute-style format
    
    Creates a response format that mimics Google's API responses with:
    - Multiple segments with length prefixes
    - Various message types (di, af.httprm, e, wrb.fr)
    - Scattered encrypted data across nested structures
    - Unicode escaping
    """
    timestamp = int(time.time() * 1000)
    micro_timestamp = timestamp * 1000 + secrets.randbelow(1000)
    
    # Generate decoy identifiers that look like Google's
    resource_id = _generate_google_id()
    request_token = _generate_google_id()
    session_token = str(secrets.randbelow(10**18))  # Large numeric string
    
    # Split the ciphertext and IV for scattering across the response structure
    ct_parts = _split_data_for_scatter(ciphertext, 3)
    iv_parts = _split_data_for_scatter(iv, 2)
    
    # Apply unicode escaping to parts (mimics Google's response format)
    ct_parts_escaped = [_apply_unicode_escaping(p) for p in ct_parts]
    iv_parts_escaped = [_apply_unicode_escaping(p) for p in iv_parts]
    
    # Create deeply nested structure similar to Google Flights response
    # Structure indices (used for extraction during decoding):
    # - inner[4][0][0][1][1]: ciphertext part 1 (in resource metadata)
    # - inner[4][0][0][6][5]: IV part 1 (in multi-carrier info)
    # - inner[7]: ciphertext part 2
    # - inner[8][2]: ciphertext part 3 (in timestamp array)
    # - inner[9]: IV part 2
    inner_data = [
        # Index 0: Resource identifier header
        [
            None, 
            None, 
            2, 
            resource_id
        ],
        None,  # Index 1: Reserved
        None,  # Index 2: Reserved
        None,  # Index 3: Reserved
        # Index 4: Nested resource entries (contains ct_part[0] and iv_part[0])
        [
            [
                [
                    _generate_resource_path(),  # Resource path like /m/0_5xc
                    # Index [1]: Resource metadata array
                    [
                        [None, secrets.randbelow(10000)],
                        ct_parts_escaped[0],  # [1][1]: First ciphertext part
                    ],
                    None,
                    None,
                    None,
                    None,
                    # Index [6]: Multi-carrier/route info array
                    [
                        "multi",
                        request_token,
                        secrets.randbelow(10),
                        secrets.randbelow(10000),
                        None,
                        iv_parts_escaped[0],  # [6][5]: First IV part
                        _generate_resource_path(),
                        None,
                        0
                    ],
                    None,
                    None,
                    True,
                    1,
                    2,
                    None,
                    None,
                    secrets.randbelow(10)
                ]
            ]
        ],
        None,  # Index 5: Reserved
        None,  # Index 6: Reserved
        _safe_get_part(ct_parts_escaped, 1),  # Index 7: Second ciphertext part
        # Index 8: Timestamp and metadata array
        [
            [micro_timestamp, secrets.randbelow(100000000), secrets.randbelow(10000000000)],
            None,
            _safe_get_part(ct_parts_escaped, 2),  # [8][2]: Third ciphertext part
            None,
            None,
            [[3]]
        ],
        _safe_get_part(iv_parts_escaped, 1)  # Index 9: Second IV part
    ]
    
    # Create the wrb.fr segment
    wrb_fr_segment = [
        ["wrb.fr", None, json.dumps(inner_data)]
    ]
    wrb_fr_str = json.dumps(wrb_fr_segment)
    wrb_fr_len = len(wrb_fr_str)
    
    # Create additional segments like Google does
    di_segment = json.dumps([
        ["di", secrets.randbelow(10000)]
    ])
    di_len = len(di_segment)
    
    httprm_segment = json.dumps([
        ["af.httprm", secrets.randbelow(10000), session_token, secrets.randbelow(100)]
    ])
    httprm_len = len(httprm_segment)
    
    e_segment = json.dumps([
        ["e", secrets.randbelow(10), None, None, secrets.randbelow(10000000)]
    ])
    e_len = len(e_segment)
    
    # Combine all segments with length prefixes (Google's format)
    response = f"{wrb_fr_len}\n{wrb_fr_str}{di_len}\n{di_segment}{httprm_len}\n{httprm_segment}{e_len}\n{e_segment}"
    
    return response


def decode_response_from_google_format(body: str) -> Optional[Tuple[str, str]]:
    """
    Decode Google BatchExecute-style response to original encrypted data.
    
    Handles the sophisticated scattered format by extracting all parts
    and reassembling the ciphertext and IV.
    """
    try:
        # Parse the multi-segment response
        segments = _parse_google_segments(body)
        
        if not segments:
            return None
        
        # Find the wrb.fr segment
        wrb_fr_data = None
        for segment in segments:
            if isinstance(segment, list) and len(segment) > 0:
                if isinstance(segment[0], list) and len(segment[0]) >= 3 and segment[0][0] == 'wrb.fr':
                    wrb_fr_data = segment[0]
                    break
        
        if not wrb_fr_data:
            return None
        
        # Parse the inner payload
        inner_str = wrb_fr_data[2]
        inner = json.loads(inner_str)
        
        if not isinstance(inner, list):
            return None
        
        # Extract scattered ciphertext and IV parts from the nested structure
        # See encode_response_to_google_format() for the structure documentation
        ct_parts = []
        iv_parts = []
        
        # Ciphertext part 1: inner[4][0][0][1][1] (in resource metadata)
        try:
            ct_parts.append(_unescape_unicode(inner[4][0][0][1][1]))
        except (IndexError, TypeError):
            pass
        
        # Ciphertext part 2: inner[7] (standalone slot)
        try:
            if inner[7] is not None:
                ct_parts.append(_unescape_unicode(inner[7]))
        except (IndexError, TypeError):
            pass
        
        # Ciphertext part 3: inner[8][2] (in timestamp array)
        try:
            if inner[8] is not None and inner[8][2] is not None:
                ct_parts.append(_unescape_unicode(inner[8][2]))
        except (IndexError, TypeError):
            pass
        
        # IV part 1: inner[4][0][0][6][5] (in multi-carrier info)
        try:
            iv_parts.append(_unescape_unicode(inner[4][0][0][6][5]))
        except (IndexError, TypeError):
            pass
        
        # IV part 2: inner[9] (standalone slot)
        try:
            if inner[9] is not None:
                iv_parts.append(_unescape_unicode(inner[9]))
        except (IndexError, TypeError):
            pass
        
        # Reassemble the scattered parts back into complete data
        ciphertext = _reassemble_scattered_data(ct_parts)
        iv = _reassemble_scattered_data(iv_parts)
        
        if ciphertext is None or iv is None:
            return None
        
        return ciphertext, iv
        
    except (json.JSONDecodeError, TypeError, KeyError, IndexError):
        return None


def _parse_google_segments(body: str) -> List[Any]:
    """
    Parse Google's multi-segment response format.
    
    Format: <length>\\n<json_segment><length>\\n<json_segment>...
    """
    segments = []
    remaining = body.strip()
    
    while remaining:
        # Find the length prefix
        newline_idx = remaining.find('\n')
        if newline_idx == -1:
            break
        
        length_str = remaining[:newline_idx]
        if not length_str.isdigit():
            break
        
        segment_length = int(length_str)
        segment_start = newline_idx + 1
        segment_end = segment_start + segment_length
        
        if segment_end > len(remaining):
            # Partial segment - try to parse what we have
            segment_str = remaining[segment_start:]
        else:
            segment_str = remaining[segment_start:segment_end]
        
        try:
            segment = json.loads(segment_str)
            segments.append(segment)
        except json.JSONDecodeError:
            pass
        
        remaining = remaining[segment_end:] if segment_end <= len(remaining) else ""
    
    return segments


def _unescape_unicode(s: str) -> str:
    """Reverse Unicode escaping"""
    if not isinstance(s, str):
        return s
    
    replacements = {
        '\\u003d': '=',
        '\\u003c': '<',
        '\\u003e': '>',
        '\\u0026': '&',
    }
    result = s
    for escape, char in replacements.items():
        result = result.replace(escape, char)
    return result


def encode_to_google_format(ciphertext: str, iv: str) -> Dict[str, str]:
    """
    Encode encrypted data into Google BatchExecute-style format (for client use)
    
    Transforms ciphertext, iv into a nested array structure
    that mimics Google's internal API format
    """
    state = get_session_state()
    timestamp = int(time.time() * 1000)
    
    # Create nested array structure similar to Google's format
    inner_payload = [
        None,
        [
            [timestamp, secrets.randbelow(1000000001), secrets.randbelow(10000000001)],
            None,
            None,
            None,
            None,
            [[1]]
        ],
        0,
        ciphertext,
        iv,
    ]
    
    wrb_fr_payload = [
        ['wrb.fr', None, json.dumps(inner_payload)]
    ]
    
    f_req = json.dumps([
        None,
        json.dumps(wrb_fr_payload)
    ])
    
    return {
        'f.req': f_req,
        'at': state.at,
    }


def parse_form_urlencoded(body: str) -> Dict[str, str]:
    """Parse URL-encoded form data"""
    result = {}
    for pair in body.split('&'):
        if '=' in pair:
            key, value = pair.split('=', 1)
            # Use unquote_plus to handle + as space (standard form encoding)
            result[urllib.parse.unquote_plus(key)] = urllib.parse.unquote_plus(value)
    return result


def get_obfuscation_headers() -> Dict[str, str]:
    """Generate Google-style response headers"""
    return {
        'x-content-type-options': 'nosniff',
        'x-frame-options': 'SAMEORIGIN',
        'x-xss-protection': '0',
        'cache-control': 'no-cache, no-store, max-age=0, must-revalidate',
        'pragma': 'no-cache',
        'content-disposition': 'attachment; filename="response.bin"; filename*=UTF-8\'\'response.bin',
    }
